homeassistant/components/hunterdouglas_powerview/button.py

```diff
@@ -0,0 +1,142 @@
+        press_action=lambda shade, button: shade.calibrate(),
+    ),
+    PowerviewButtonDescription(
+        key="identify",
+        name="Identify",
+        icon="mdi:crosshairs-question",
+        device_class=ButtonDeviceClass.UPDATE,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        press_action=lambda shade, button: shade.jog(),
+    ),
+    PowerviewButtonDescription(
+        key="update",
+        name="Force Update",
+        icon="mdi:autorenew",
+        device_class=ButtonDeviceClass.UPDATE,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        press_action=lambda shade, button: button.async_force_refresh(),
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set buttons for device."""
+
+    pv_data = hass.data[DOMAIN][entry.entry_id]
+    room_data: dict[str | int, Any] = pv_data[PV_ROOM_DATA]
+    shade_data = pv_data[PV_SHADE_DATA]
+    pv_request = pv_data[PV_API]
+    coordinator: PowerviewShadeUpdateCoordinator = pv_data[COORDINATOR]
+    device_info: dict[str, Any] = pv_data[DEVICE_INFO]
+
+    entities: list[ButtonEntity] = []
+    for raw_shade in shade_data.values():
+        shade: BaseShade = PvShade(raw_shade, pv_request)
+        name_before_refresh = shade.name
+        room_id = shade.raw_data.get(ROOM_ID_IN_SHADE)
+        room_name = room_data.get(room_id, {}).get(ROOM_NAME_UNICODE, "")
+
+        for description in BUTTONS:
+            entities.append(
+                PowerviewButton(
+                    coordinator,
+                    device_info,
+                    room_name,
+                    shade,
+                    name_before_refresh,
+                    description,
+                )
+            )
+
+    async_add_entities(entities)
+
+
+class PowerviewButton(ShadeEntity, ButtonEntity):
+    """Representation of an advanced feature button."""
+
+    def __init__(
+        self,
+        coordinator: PowerviewShadeUpdateCoordinator,
+        device_info,
+        room_name,
+        shade,
+        name,
+        description: PowerviewButtonDescription,
+    ):
+        """Initialize the button entity."""
+        super().__init__(coordinator, device_info, room_name, shade, name)
+        self.entity_description: PowerviewButtonDescription = description
+        self._attr_name = f"{self._shade_name} {description.name}"
+        self._attr_unique_id = f"{self._attr_unique_id}_{description.key}"
+
+    async def async_press(self) -> None:
+        """Handle the button press."""
+        await self.entity_description.press_action(self._shade, self)
+
+    async def async_force_refresh(self) -> None:
+        """Refresh shade position."""
+        _LOGGER.debug("Manual update of shade data run for %s", self._shade_name)
+        await self._shade.refresh()
+        self.data.update_shade_positions(self._shade.raw_data)
+        self.async_write_ha_state()
```

Comment:
This seems like it duplicates the functionality of https://www.home-assistant.io/integrations/homeassistant/#service-homeassistantupdate_entity

If you rename this to async_update and remove `self.async_write_ha_state()` and put in the base shade entity this should make all entities be able to do force update with the built-in call

Reply:
i've just pushed a change to do that - and made the change to reflect in cover.py aswell

Could you please test the button for me, it is functioining the same with this code as it was before but still makes me question - the force update button definitely updates the shade position (verified by looking at the shade position in a browser window and seeing the position update) but the actual slider in HA doesnt update immediately - takes around 30 seconds for it to change

Reply:
i put some dummy logging events in for testing and can definitely see it being put into the coordinator so wasnt sure why the UI would be lagging behind

Reply:
I thought you said to remove a `self._async_write_ha_state()` call.
In particlular since this replaces the force refresh in the cover.py what code do i need to add to make sure it writes the state back to HA too ?

Are you saying i need to add that back to the end of the async_update call defined here ?
https://github.com/home-assistant/core/pull/73061/commits/1ad555b141ad001ea2769865f56ce01de947d1c4

Reply:
> Are you saying i need to add that back to the end of the async_update call defined here ?
> [1ad555b](https://github.com/home-assistant/core/commit/1ad555b141ad001ea2769865f56ce01de947d1c4)

If you call it from the service, the service will do it for you.  If you call it from the button you would still need it since it doesn't do that for you..